### PR TITLE
Counter/defaultdict: Key type 'str' if initialized with keywords

### PR DESCRIPTION
**Pull request by srittau**
Closes: #7045

### CODE REVIEW

**Review by AlexWaygood — COMMENTED** [review]
I think the `defaultdict` overloads on lines 309, 314 and 320 might also need to be altered.

**Comment by srittau** [comment]
Good catch. In the latter two cases, I enforced a `str` type for the iterable/mapping. While in theory you could do `defaultdict([(123, 456)], foo=999)` to get a `defaultdict[str | int, int]`, I think it's safer to assume that this is an error.

**Comment by AlexWaygood** [comment]
> While in theory you could do `defaultdict([(123, 456)], foo=999)` to get a `defaultdict[str | int, int]`, I think it's safer to assume that this is an error.

I agree. Unless mypy_primer says otherwise, I think it's safe to assume `defaultdict[str | int, int]` is pretty rare (and you can always just `# type: ignore` in the rare case where that is actually what you need).

**Comment by AlexWaygood** [comment]
I have no idea what pandas is doing there, but it looks like pretty dynamic code, and [they already have a tonne of `type: ignore`s on that line](https://github.com/pandas-dev/pandas/blob/20d096408afbdf0ca18638b88fb34026b5498ba7/pandas/io/parsers/base_parser.py#L1029). So I'm not sure it's worth worrying about.

File: stdlib/collections/__init__.pyi [hunk 1]
```diff
@@ -298,22 +300,30 @@ class OrderedDict(dict[_KT, _VT], Reversible[_KT], Generic[_KT, _VT]):
 class defaultdict(dict[_KT, _VT], Generic[_KT, _VT]):
     default_factory: Callable[[], _VT] | None
     @overload
-    def __init__(self, **kwargs: _VT) -> None: ...
+    def __init__(self: defaultdict[_KT, _VT]) -> None: ...
```
Comment (Akuli) [inline stdlib/collections/__init__.pyi:303]: Why does this overload have an explicit self type?

Comment (srittau) [reply stdlib/collections/__init__.pyi:303]: For consistency with `dict`.

**Review by Akuli — APPROVED** [review]

**Pull request merged**